### ISSUE TITLE
False positive in overload checking: `type "(**P@f) -> R@f" cannot be assigned to type "(**P@f) -> R@f"`

### ISSUE THREAD

**Issue by rsokl** [microsoft/pyright#5416]
On pyright 1.1.316 the following overloaded function is marked with an error that I believe is a false-positive:

```
 error: Overloaded implementation is not consistent with signature of overload 2
    Function return type "((**P2@f) -> R2@f) -> A[P2@f, R2@f]" is incompatible with type "((((**P2@f) -> R2@f)) -> A[P2@f, R2@f]) | A[P@f, R@f]"
      Type "((**P2@f) -> R2@f) -> A[P2@f, R2@f]" cannot be assigned to type "((((**P2@f) -> R2@f)) -> A[P2@f, R2@f]) | A[P@f, R@f]"
        Type "((**P2@f) -> R2@f) -> A[P2@f, R2@f]" cannot be assigned to type "((**P2@f) -> R2@f) -> A[P2@f, R2@f]"
          Parameter 1: type "(**P2@f) -> R2@f" cannot be assigned to type "(**P2@f) -> R2@f"
            Type "(**P2@f) -> R2@f" cannot be assigned to type "(**P2@f) -> R2@f"
        "function" is incompatible with "A[P@f, R@f]" (reportGeneralTypeIssues)
```

```python
from typing import Generic, TypeVar, overload, cast, Union
from typing_extensions import ParamSpec

R = TypeVar("R")
P = ParamSpec("P")
P2 = ParamSpec("P2")
R2 = TypeVar("R2")

class A(Generic[P, R]):
    def __init__(self, __func: Callable[P, R]):
        ...

@overload
def f(
    __func: Callable[P, R],
    *,
    Wrapper: Type[A[Any, Any]] = ...,
) -> A[P, R]:
    ...

@overload
def f(
    __func: Literal[None] = None,
    *,
    Wrapper: Type[A[Any, Any]] = ...,
) -> Callable[[Callable[P2, R2]], A[P2, R2]]:
    ...

def f(
    __func: Optional[Callable[P, R]] = None,
    *,
    Wrapper: Type[A[Any, Any]] = A,
) -> Union[Callable[[Callable[P2, R2]], A[P2, R2]], A[P, R]]:
    if __func is not None:
        return cast(A[P, R], Wrapper(__func))

    def wrap(f: Callable[P2, R2]) -> A[P2, R2]:
        return cast(A[P2, R2], Wrapper(f))

    return wrap
```

**Comment by rsokl** [comment]
Here is a simpler example that raises the same error:

```python
class A(Generic[P, R]):
    def __init__(self, __func: Callable[P, R]):
        ...


@overload
def f(__func: int) -> int:
    ...

@overload
def f(__func: Literal[None] = None) -> Callable[[Callable[P, R]], A[P, R]]:
    ...

def f(
    __func: Optional[int] = None,
) -> Union[Callable[[Callable[P, R]], A[P, R]], int]:
    ...

```

**Comment by GugelRobin** [comment]
Related to (or duplicate of) #5384

**Comment by erictraut** [comment]
Thanks for reporting. Yes, this turns out to be the same underlying issue as #5384. It will be addressed in the next release.
